### PR TITLE
Initialize tracks on device

### PR DESCRIPTION
**Pull request by amandalund**
Add some of the functionality for initializing tracks on device from primaries and secondaries in a way that is reproducible.

### CODE REVIEW

File: src/sim/VacancyStore.hh [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+#include "base/DeviceVector.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Manage device data for vacancies.
+ */
+class VacancyStore
```
Comment (sethrj) [inline src/sim/VacancyStore.hh:18]: Do you anticipate the `VacancyStore` being anything but a detail-level class? For helper data like this, it might be easier for the helper class that calls the `InitializeTracks` functions to just store the `DeviceVector` internally. Your call of course.

Also, since I see the "shrinkable device vector" paradigm is used more than once here, it might be worth moving into the DeviceVector class, although the caveat is that it should only be used for trivially destructible objects.

File: test/sim/InitializeTracks.test.cc [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file InitializeTracks.test.cc
+//---------------------------------------------------------------------------//
+#include "InitializeTracks.test.hh"
+
+#include "physics/base/SecondaryAllocatorStore.hh"
+#include "gtest/Main.hh"
+#include "gtest/Test.hh"
+#include "sim/InitializeTracks.hh"
```
Comment (sethrj) [inline test/sim/InitializeTracks.test.cc:13]: To guard against missing includes in the source file, it's best to put the header for the class being tested at the very top of the file.

File: test/sim/InitializeTracks.test.hh [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+{
+using namespace celeritas;
+
+//---------------------------------------------------------------------------//
+// TESTING INTERFACE
+//---------------------------------------------------------------------------//
+//! Interactor
+struct Interactor
+{
+    __device__ Interactor(const ParticleTrackView& particle,
```
Comment (sethrj) [inline test/sim/InitializeTracks.test.hh:29]: This is being included from a `.cc` file, so I'm surprised the line compiles -- one of the other headers above must have a CUDA APi include that you're indirectly including. You can put this in a `.cuh`, move it into the `.test.cu` file, or change the decorators to `CELER_FUNCTION`.

**Review by sethrj — CHANGES_REQUESTED** [review]
First round of feedback. The core functionality looks good, but the way all the classes integrate is still kind of fuzzy (by necessity, because we don't yet have a high-level plan for how the pieces fit together). How about we talk on Monday's infrastructure meeting about converging on a high-level starting point plan that we can adapt as time goes on?

File: src/sim/InitializeTracks.cu [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+                         const ParticleStatePointers  pstates,
+                         const GeoParamsPointers      gparams,
+                         const GeoStatePointers       gstates)
+{
+    auto thread_id = KernelParamCalculator::thread_id().get();
+    if (thread_id < num_tracks)
+    {
+        // Get the track initializer, starting from the back of the vector
+        size_type               init_id = initializers.size() - thread_id - 1;
+        const TrackInitializer& init    = initializers.data()[init_id];
```
[the remark this answers is not in the document]
Comment (amandalund) [reply src/sim/InitializeTracks.cu:44]: Oops, right.

File: src/sim/InitializeTracks.cu [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+        // Determine which indices in the track states are available for
+        // initializing new particles
+        if (action_killed(result.action))
+        {
+            vacancies.data()[thread_id] = thread_id;
+        }
+        else
+        {
+            // Flag as a track that's still alive
+            vacancies.data()[thread_id] = interactions.size();
```
[the remark this answers is not in the document]
Comment (amandalund) [reply src/sim/InitializeTracks.cu:80]: I like that, much clearer.

File: src/sim/TrackInitializerStore.hh [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    TrackInitializerStore(size_type capacity);
+
+    // Get the number of elements
+    size_type capacity() const { return allocation_.size(); }
+
+    // Get the number of elements
+    size_type size() const { return size_; }
+
+    // Change the size (without changing capacity)
+    void resize(size_type size) { size_ = size; }
```
[the remark this answers is not in the document]
Comment (amandalund) [reply src/sim/TrackInitializerStore.hh:39]: Absolutely right.

File: src/sim/VacancyStore.hh (hunk 1 above)
Comment (amandalund) [reply src/sim/VacancyStore.hh:18]: No, this was more of a placeholder since I didn't have the class to call those functions and manage that data yet, but ultimately storing it in that helper class makes the most sense.

File: test/sim/InitializeTracks.test.hh (hunk 3 above)
Comment (amandalund) [reply test/sim/InitializeTracks.test.hh:29]: Yikes, good catch.

File: src/sim/InitializeTracks.cu [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+                         const ParticleParamsPointers pparams,
+                         const ParticleStatePointers  pstates,
+                         const GeoParamsPointers      gparams,
+                         const GeoStatePointers       gstates)
+{
+    auto thread_id = KernelParamCalculator::thread_id().get();
+    if (thread_id < num_tracks)
+    {
+        // Get the track initializer, starting from the back of the vector
+        size_type               init_id = initializers.size() - thread_id - 1;
```
[the remark this answers is not in the document]
Comment (amandalund) [reply src/sim/InitializeTracks.cu:43]: You're right, that `num_tracks` is superfluous. The `vacancies` and `initializers` aren't necessarily the same size; `vacancies` is resized just before this kernel launch so it will have at most the same number of elements as `initializers`, but `initializers` could be larger. Then `n = vacancies.size()` track initializers are pulled off the back of that vector so that the first vacancy is filled with the last track initializer, the second vacancy with the next-to-last, etc. This is all pretty poorly documented (I confused myself looking back at it just now), so I’ll go back and add some more descriptive comments.

**Comment by amandalund** [comment]
Thanks for looking this over! It’s definitely still pretty rough and missing some pieces — I was hoping we could go over some things before I went much further with it, so that sounds great.

File: src/sim/InitializeTracks.cu (hunk 7 above)
[the remark this answers is not in the document]
Comment (sethrj) [reply src/sim/InitializeTracks.cu:42]: Hehe there's a lot to keep "track" of. 😉 Thanks!

**Comment by sethrj** [comment]
> Thanks for looking this over! It’s definitely still pretty rough and missing some pieces — I was hoping we could go over some things before I went much further with it, so that sounds great.

@amandalund For sure, and thanks for the hard work! If Monday's infrastructure meeting is too late (or if you want to do a one-on-one rather than have to have other people watching) I'm free from noon to 3ish eastern time.

**Comment by sethrj** [comment]
@amandalund This is looking great. I had to make a few changes (to fix the build when CUDA is disabled, I wrote stubs for the curand code) that I can move to another branch, and I did a tiny bit of restructuring (moving files to a "detail" namespace). Should I just push those changes on top of your branch?

**Comment by amandalund** [comment]
Awesome, thanks @sethrj! You can just push them here.

File: src/sim/detail/InitializeTracks.cu [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+
+        SimTrackView sim(states.sim, ThreadId(thread_id));
+
+        // The track is alive: mark this track slot as active
+        if (sim.alive())
+        {
+            inits.vacancies[thread_id] = flag_alive();
+        }
+        // The track is dead and produced secondaries: fill the empty track
+        // slot with the first secondary and mark the track slot as active
```
Comment (sethrj) [inline src/sim/detail/InitializeTracks.cu:118]: Instead of adding visual separation between the `if` block and its sister `else`, move the comments in both cases inside the appropriate conditionals.

File: src/sim/detail/InitializeTracks.cu [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+            if (secondary)
+            {
+                TrackInitializer& init
+                    = inits.initializers[inits.initializers.size() + offset_id];
+
+                // Store the thread ID of the secondary's parent
+                inits.parent[offset_id] = thread_id;
+
+                // Calculate the track ID of the secondary
+                unsigned int track_id = 1 + inits.track_count + offset_id++;
```
Comment (sethrj) [inline src/sim/detail/InitializeTracks.cu:196]: Is the `1 +` here correct? For the first secondary of the first primary, I would expect the `TrackId` to be 1 (with a parent ID of 0), but here it would be 2 I think.

Digging around and trying to understand the `track_count` made me realize something else -- we really need independent track counters for each separate *event* in flight because right now the `TrackId` is effectively a *global*  counter, isn't it? Can @pcanal confirm whether the track ID should be on a per-event basis?

File: src/sim/detail/InitializeTracks.cu [hunk 10]
```diff
@@ -0,0 +1,339 @@
...
+        thrust::plus<size_type>());
+
+    CELER_CUDA_CALL(cudaDeviceSynchronize());
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate the exclusive prefix sum of the number of surviving secondaries
+ * from each interaction.
```
Comment (sethrj) [inline src/sim/detail/InitializeTracks.cu:324]: Can you add a line explaining "exclusive prefix sum"? I think it's doing the same as `numpy.cumsum(a, out=a)`?

**Review by sethrj — CHANGES_REQUESTED** [review]
Amanda, this was a huge lift and looks excellent. The higher-level data organization is still a little muddled (as much my fault as it is yours) so let's talk Monday about how to organize the different primary/secondary/track states.

File: src/base/detail/InitializedValue.hh [hunk 11]
```diff
@@ -43,7 +43,11 @@ class InitializedValue
     }
 
     //! Implicit assign from type
-    InitializedValue operator=(const T& value) { value_ = value; }
+    InitializedValue operator=(const T& value)
+    {
+        value_ = value;
+        return *this;
+    }
```
[the remark this answers is not in the document]
Comment (amandalund) [reply src/base/detail/InitializedValue.hh:50]: Ah you're right, I missed that.

File: src/geometry/GeoTrackView.i.hh [hunk 12]
```diff
@@ -60,6 +60,25 @@ CELER_FUNCTION GeoTrackView& GeoTrackView::operator=(const Initializer_t& init)
     return *this;
 }
 
+//---------------------------------------------------------------------------//
+/*!
+ * Construct the state from a direction and a copy of the parent state.
+ */
+CELER_FUNCTION void
+GeoTrackView::copy_state(const GeoTrackView& parent, const Real3 direction)
```
[the remark this answers is not in the document]
Comment (amandalund) [reply src/geometry/GeoTrackView.i.hh:68]: Good idea, will do.

File: src/sim/detail/InitializeTracks.cu (hunk 10 above)
Comment (amandalund) [reply src/sim/detail/InitializeTracks.cu:324]: Yes, I'll add a better explanation of what this is doing. This would be equivalent to `np.cumsum(np.concatenate(([0], a[:-1])), out=a)`, so if the input were `a = [2, 4, 1, 0, 3]`, the result would be `a = [0, 2, 6, 7, 7]`. I.e., when calculating the output at index `i`, the `i`'th element isn't included in the sum.

File: src/io/EventReader.cc [hunk 13]
```diff
@@ -47,6 +47,9 @@ EventReader::result_type EventReader::operator()()
         }
         ++event_id;
 
+        // Track numbering starts at 1 since 0 is reserved for the parent ID
+        int track_id = 1;
```
[the remark this answers is not in the document]
Comment (amandalund) [reply src/io/EventReader.cc:51]: I'm probably misunderstanding how the Geant track numbering works (wouldn't be the first time).. I thought primary particles had parent ID zero, which is why I had primary track IDs starting from 1, but I'm definitely not sure that's right.

File: src/sim/detail/InitializeTracks.cu (hunk 9 above)
Comment (amandalund) [reply src/sim/detail/InitializeTracks.cu:196]: You're right, we need a separate track counter for each event. The `1 +` comes from numbering the primary track IDs starting at 1 instead of 0, but there are a couple other places where I'm not assigning the track ID correctly yet for secondaries... let me fix up all the track numbering.

File: src/sim/TrackInitializerStore.cc [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+
+    ENSURE(result);
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Create track initializers on device from primary particles.
+ */
+void TrackInitializerStore::create_from_primaries(span<const Primary> primaries)
```
[the remark this answers is not in the document]
Comment (amandalund) [reply src/sim/TrackInitializerStore.cc:58]: I like it. Any other naming suggestions are welcome 🙂

File: src/sim/TrackInitializerStore.cc [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+{
+    REQUIRE(primaries.size()
+            <= initializers_.capacity() - initializers_.size());
```
[the remark this answers is not in the document]
Comment (amandalund) [reply src/sim/TrackInitializerStore.cc:61]: Good point, I'll rephrase it.

File: src/io/EventReader.cc (hunk 13 above)
[the remark this answers is not in the document]
Comment (sethrj) [reply src/io/EventReader.cc:51]: Oh, you know what I could be misunderstanding as well, I didn't even think of what the *parent* IDs for the "primary" particles are. I guess I assumed they'd be unassigned instead of having a meaningful value.

I'm not sure what exactly the Geant family of codes does, but I think the most sensible thing for us to do is to use track IDs that correspond to the individual tracks that we transport -- 0 through `num_primaries - 1` for the primaries for a particular event. The parent IDs for those I would leave with the "unassigned" ID.

[the remark this answers is not in the document]
Comment (amandalund) [reply src/io/EventReader.cc:51]: Works for me!

File: src/sim/TrackInitializer.hh [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+/*!
+ * Lightweight version of a track used to initialize new tracks from primaries
+ * or secondaries
+ */
+struct TrackInitializer
+{
+    // Simulation state
+    SimTrackView::Initializer_t      sim;
+    GeoTrackView::Initializer_t      geo;
+    ParticleTrackView::Initializer_t particle;
```
[the remark this answers is not in the document]
Comment (amandalund) [reply src/sim/TrackInitializer.hh:30]: Oh my bad, you already had the initializers defined in the `Pointers` files so no need to go through the view. Fixed, and moved this into `TrackInitializerPointers.hh`.

File: src/sim/detail/InitializeTracks.cu [hunk 17]
```diff
@@ -0,0 +1,339 @@
...
+ */
+__global__ void locate_alive_kernel(const StatePointers            states,
+                                    const ParamPointers            params,
+                                    const TrackInitializerPointers inits)
+{
+    auto thread_id = KernelParamCalculator::thread_id().get();
+    if (thread_id < states.size())
+    {
+        // Secondary to copy to the parent's track slot if the parent has died
+        Secondary firstborn = Secondary::from_cutoff();
```
[the remark this answers is not in the document]
Comment (amandalund) [reply src/sim/detail/InitializeTracks.cu:94]: Agreed, I think that would make more sense.

File: src/sim/detail/InitializeTracks.cu [hunk 18]
```diff
@@ -0,0 +1,339 @@
...
+__global__ void init_tracks_kernel(const StatePointers            states,
+                                   const ParamPointers            params,
+                                   const TrackInitializerPointers inits)
+{
+    auto thread_id = KernelParamCalculator::thread_id().get();
+    if (thread_id < inits.vacancies.size())
+    {
+        // Get the track initializer, starting from the back of the vector
+        const TrackInitializer& init
+            = inits.initializers[inits.initializers.size() - thread_id - 1];
```
[the remark this answers is not in the document]
Comment (amandalund) [reply src/sim/detail/InitializeTracks.cu:50]: Sure thing. Not exactly -- I think think the only case where we need to worry about the parent state being overwritten is when the parent track produces secondaries and is killed (if it doesn't produce secondaries no other track will need the state, and if it's alive nothing will be initialized there). This case is handled in the kernel that finds the vacancies and counts the secondaries: one of the secondaries is immediately copied into the dead track (preserving the geometry state) and the track is marked as alive. So when we get to this point, it should be safe to copy any of the states in any order.

Track initializers are pulled from the back of the vector because those are the most recently added (and therefore the ones that still might have a parent they can copy the geometry state from) and because there could be more initializers than empty track slots.

[the remark this answers is not in the document]
Comment (sethrj) [reply src/sim/detail/InitializeTracks.cu:50]: Great! Thanks for the explanation.

File: src/io/EventReader.cc (hunk 13 above)
[the remark this answers is not in the document]
Comment (sethrj) [reply src/io/EventReader.cc:51]: Per our discussion, Geant4 particle guns start with track ID of 1 as the initial primary, and it has a parent of 0. (This is actually what your first implementation said.) But for events read in externally they can basically have an arbitrary ID. But we are going to do our own thing for now, and make sure that (when needed) we can store the metadata that associates our internal track IDs to HEPEVT-provided IDs.

**Comment by sethrj** [comment]
add to whitelist

File: src/sim/detail/InitializeTracks.cu [hunk 19]
```diff
@@ -0,0 +1,361 @@
...
+/*!
+ * Find empty slots in the track vector and count the number of secondaries
+ * that survived cutoffs for each interaction. If the track is dead and
+ * produced secondaries, fill the empty track slot with one of the secondaries.
+ */
+__global__ void locate_alive_kernel(const StatePointers            states,
+                                    const ParamPointers            params,
+                                    const TrackInitializerPointers inits)
+{
+    auto thread_id = KernelParamCalculator::thread_id().get();
```
Comment (sethrj) [inline src/sim/detail/InitializeTracks.cu:99]: In a case of "six of one, half a dozen of the other", it might be better to leave the `ThreadId` in its original type here:
```suggestion
    const auto thread_id = KernelParamCalculator::thread_id();
```
The `if` will still work, and below you'll just replace `thread_id` with `thread_id.get()`; and `ThreadId(thread_id)` with `thread_id`.

I think it's a bit "safer" this way because the type safety is only removed locally to where it's needed; and it's not much more verbose because of the number of places where the typed thread ID is actually used.

File: src/sim/TrackInitializerStore.cc [hunk 20]
```diff
@@ -0,0 +1,225 @@
...
+    {
+        const auto event_id = primary.event_id.get();
+        if (host_track_counter.size() <= event_id)
+        {
+            host_track_counter.resize(event_id + 1);
+        }
+        ++host_track_counter[event_id];
+    }
+    track_counter_ = DeviceVector<ull_int>(host_track_counter.size());
+    track_counter_.copy_to_device(make_span(host_track_counter));
```
Comment (sethrj) [inline src/sim/TrackInitializerStore.cc:52]: This is an excellent "temporary" workaround to the event-specific thread ID that we'll put in a new MR for.

**Review by sethrj — CHANGES_REQUESTED** [review]
A couple more minor things sorry! 😔 It's looking really good though 😄 and these should be quick fixes.

File: src/CMakeLists.txt [hunk 21]
```diff
@@ -81,17 +87,27 @@ if(CELERITAS_USE_VecGeom)
...
 
 add_library(celeritas ${SOURCES})
 add_library(Celeritas::Core ALIAS celeritas)
+
+if(CELERITAS_USE_VecGeom AND CELERITAS_USE_CUDA)
+  set_target_properties(celeritas PROPERTIES
+    LINKER_LANGUAGE CUDA
+    CUDA_SEPARABLE_COMPILATION ON
+  )
+endif()
```
[the remark this answers is not in the document]
Comment (amandalund) [reply src/CMakeLists.txt:109]: Oh hmm it needs to be after we add the library, right? But I could rearrange things (maybe move `add_library` to the top and use `target_sources`?) if we want to keep these together.

File: src/sim/TrackInitializerStore.hh [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+    //!
+    using constVecPrimary = const std::vector<Primary>;
+    //@}
+
+  public:
+    // Construct with the number of tracks, the maximum number of track
+    // initializers to store on device, and the primary particles
+    explicit TrackInitializerStore(size_type        num_tracks,
+                                   size_type        capacity,
+                                   constVecPrimary& primaries);
```
[the remark this answers is not in the document]
Comment (amandalund) [reply src/sim/TrackInitializerStore.hh:36]: Oh no you're totally right, I'll change this.

File: src/sim/detail/InitializeTracks.cu (hunk 19 above)
Comment (amandalund) [reply src/sim/detail/InitializeTracks.cu:99]: Fair enough!

File: src/CMakeLists.txt (hunk 21 above)
[the remark this answers is not in the document]
Comment (sethrj) [reply src/CMakeLists.txt:113]: Oh right. Well just leave it as is then! 🙂

**Review by sethrj — APPROVED** [review]

**Comment by sethrj** [comment]
Related to #119 @pcanal : the reason @amandalund had to add the separable compilation linker flags to the main celeritas library (in src/CMakeLists) was she hit the following build error:
```
[1/56] Building CUDA object src/CMakeFiles/celeritas.dir/sim/InitializeTracks.cu.o
FAILED: src/CMakeFiles/celeritas.dir/sim/InitializeTracks.cu.o 
/usr/local/cuda-10.2/bin/nvcc -forward-unknown-to-host-compiler -ccbin=/usr/bin/g++ -DVECCORE_ENABLE_CUDA -Dceleritas_EXPORTS -I/projects/spack/opt/spack/gcc-8.3.1/clhep/ual4i4i/lib/CLHEP-2.4.1.3/../../include -I../src -Isrc -isystem=/projects/spack/var/spack/environments/celeritas/.spack-env/view/include -isystem=/usr/local/cuda-10.2/include -isystem=/projects/spack/opt/spack/gcc-8.3.1/xerces-c/ffb3b3s/include -g -G --generate-code=arch=compute_35,code=[compute_35,sm_35] -Xcompiler=-fPIC -std=c++14 -MD -MT src/CMakeFiles/celeritas.dir/sim/InitializeTracks.cu.o -MF src/CMakeFiles/celeritas.dir/sim/InitializeTracks.cu.o.d -x cu -c ../src/sim/InitializeTracks.cu -o src/CMakeFiles/celeritas.dir/sim/InitializeTracks.cu.o
ptxas fatal   : Unresolved extern function '_ZN7vecgeom20globaldevicegeomdata25GetCompactPlacedVolBufferEv'
```

I told her I'd seen a nearly identical error when writing the demo rasterizer the first time; my solution had been to add
```
  set_target_properties(demo_rasterizer_cuda PROPERTIES
    LINKER_LANGUAGE CUDA
    CUDA_SEPARABLE_COMPILATION ON
    POSITION_INDEPENDENT_CODE ON
  )
```

**Comment by pcanal** [comment]
So the difference between the CI and my environment turned out to be ... the version of cmake.

I was using v3.15.4 which does not support yet the flag CMAKE_CUDA_ARCHITECTURES (Noticed that to the warning from cmake "this variable was set but not used").

When I tried with cmake v3.19.3 without adding that flag, the build failed (missing symbol).  This is likely due to the fact that this version of cmake no longer propagate the `-arch=sm_70` set by (I think VecGeom) and instead use default value (which turns out to be sm_52).   Once I added  `-DCMAKE_CUDA_ARCHITECTURES:STRING=70`, the build succeeded and the process succeeds ....

Bottom line, we need to require a version of cmake "greater than" v3.15.4.  v3.19.3 and which ever version the CI is using seems to be acceptable version.

**Comment by sethrj** [comment]
Took me a second, but this makes sense -- with a "default value" (where does that come from??) of `sm_52` the kernels that *don't* call vecgeom are fine, since cuda recompiles for the higher architecture at runtime based on the PTX. For the kernels that link against vecgeom I guess separable compilation gets in the way and causes the two different sets of flags to misbehave.

**Comment by pcanal** [comment]
> but are you sure the cuda flags will propagate correctly on older versions of cmake

It only *seems* to since in the end the build is not correct (see the "invalid kernel" error).

> or that the build works on newer versions of cmake if the architecture flags aren't set consistently between vecgeom and celeritas? 

Even with cmake 3.19, if the flags are inconsistent it wont work ... either visibly (missing symbol) or plausibly (but did not observe that) the same way it fails with cmake 3.15.

What I think we need is:
```
if((CMAKE_VERSION VERSION_LESS 3.18))
  # error
else()
  somehow check that the request architecture (either CMAKE_CUDA_ARCHITECTURES or whereever my sm_52 came from) match the one from VecGeom ... and if not error out.
endif()
```

**Comment by mrguilima** [comment]
Just upgrading CMake to 3.19.3 was not sufficient, my CMake run fails:
```
CMake Deprecation Warning at external/googletest/CMakeLists.txt:4 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.
  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.
```
How should googletest be upgraded?

**Comment by sethrj** [comment]
@mrguilima That's a warning, not an error. I'll see if that version can be updated. An easy way to avoid the error is to preinstall googletest and link to it externally by providing it in your cmake prefix path.
